lib/compress/zstd_compress_internal.h

```diff
@@ -220,6 +220,95 @@ struct ZSTD_CCtx_params_s {
+ * - These different internal datastructures have different setup requirements.
+ *   Some (e.g., the window buffer) don't care, and are happy to accept
+ *   uninitialized memory. Others (e.g., the matchstate tables) can accept
+ *   memory filled with unknown but bounded values (i.e., a memory area whose 
+ *   values are known to be constrained between 0 and some upper bound). If
+ *   that constraint isn't known to be satisfied, the area has to be cleared.
+ *
+ * - We would like to reuse the objects in the workspace for multiple
+ *   compressions without having to perform any expensive reallocation or
+ *   reinitialization work.
+ *
+ * - We would like to be able to efficiently reuse the workspace across
+ *   multiple compressions **even when the compression parameters change** and
+ *   we need to resize some of the objects (where possible).
+ *
+ * Workspace Layout:
+ *
+ * [                        ... workspace ...                         ]
+ * [objects][tables ... ->] free space [<- ... aligned][<- ... buffers]
+ *
+ * In order to accomplish this, the various objects that live in the workspace
+ * are divided into the following categories:
+ *
+ * - Static objects: this is optionally the enclosing ZSTD_CCtx or ZSTD_CDict,
+ *   so that literally everything fits in a single buffer. Note: if present,
+ *   this must be the first object in the workspace, since ZSTD_free{CCtx,
+ *   CDict}() rely on a pointer comparison to see whether one or two frees are
+ *   required.
+ *
+ * - Fixed size objects: these are fixed-size, fixed-count objects that are
+ *   nonetheless "dynamically" allocated in the workspace so that we can
+ *   control how they're initialized separately from the broader ZSTD_CCtx.
+ *   Examples:
+ *   - Entropy Workspace
+ *   - 2 x ZSTD_compressedBlockState_t
+ *   - CDict dictionary contents sometimes??? // TODO
+ *
+ * - Tables: these are any of several different datastructures (hash tables,
+ *   chain tables, binary trees) that all respect a common format: they are
+ *   uint32_t arrays, all of whose values are between 0 and (nextSrc - base).
+ *   Their sizes depend on the cparams.
+ *
+ * - Aligned: these buffers are used for various purposes that don't require
+ *   any initialization before they're used.
+ *
+ * - Uninitialized memory: these buffers are used for various purposes that
+ *   don't require any initialization before they're used. This means they can
+ *   be moved around at no cost for a new compression.
+ *
+ * Allocating Memory:
+ *
+ * The various types of objects must be allocated in order, so they can be
+ * correctly packed into the workspace buffer. That order is:
+ *
+ * 1. Objects
+ * 2. Buffers
+ * 3. Aligned
+ * 4. Tables
+ */
+typedef struct {
+    void* workspace;
+    void* workspaceEnd;
+
+    void* objectEnd;
+
+    void* tableEnd;
+
+    void* allocStart;
+
+    int allocFailed;
+    int workspaceOversizedDuration;
+    ZSTD_workspace_alloc_phase_e phase;
+} ZSTD_CCtx_workspace;
```

Comment:
Naming : 
`ZSTD_CCtx_workspace` seems to mean that this object is the actual `workspace` of `CCtx`.

However, reading its declaration, it's a different thing : 
it contains pointers inside the actual `workspace`,
making it possible to track its different regions.

So it's related, but different.
I suggest changing the type name to something more specific, as the current name can be misleading when the declaration is not known (which is quite possible, since there is a large distance between declaration and usage of the type).

Possible candidates : 
- `ZSTD_CCtx_workspace_descriptor` 
- `ZSTD_CCtx_workspace_layout` 
- `ZSTD_CCtx_wksp_desc`
- etc.

Reply:
Btw, is `CCtx` meaningful ?
It seems `CDict` uses the same object too ?